QgsFeatureSource::empty() method

m-kuhn opened the pull request:
The main reason is

 * It looks nice
 * It's more performant than checking `source.count() == 0`
 * It also works reliably on pg layers with estimated metadata enabled

m-kuhn committed:
Add empty() and emptyUnknown() to QgsFeatureSource

m-kuhn committed:
Add optimized versions of empty() to QgsVectorDataProvider

m-kuhn committed:
Add optimized version of QgsPostgresProvider::empty()

m-kuhn committed:
Add test for empty() to provider test suite

m-kuhn commented:
Ideas for more tests still welcome ;)

m-kuhn committed:
Fix typo

nyalldawson commented on src/core/qgsvectordataprovider.h:
```diff
@@ -160,6 +160,21 @@ class CORE_EXPORT QgsVectorDataProvider : public QgsDataProvider, public QgsFeat
+     *
+     * \since QGIS 3.2
+     */
+    virtual bool empty() const override;
```
Can you please implement these in QgsProcessingFeatureSource (proxying off to the underlying source to ensure any optimised overrides are used correctly?)

nyalldawson commented on src/core/qgsfeaturesource.h:
```diff
@@ -85,6 +85,27 @@ class CORE_EXPORT QgsFeatureSource
+     *
+     * \since QGIS 3.2
+     */
+    virtual bool emptyUnknown() const;
```
I don't like this... (sorry to be blunt !) 

Could we instead remove this and do something like:
- check featureCount(), if !=-1 then we have the answer
- if -1, then do the feature request with a limit of 1 and no geometry and no attributes and see what we get?

Then we'd have a empty() method which can always be relied on without additional checks

m-kuhn replied:
One of the main advantages of `empty()` over `featureCount()` is that it's fast. If the prerequisite to know if we can use the fast method is a call to a slow method that's rendering it a lot worse.

I wonder what should be done for a `QgsProcessingFeatureSource` for a filter algorithm. We don't know if it will be returning any features. And iterating over 1000000 features to find out it's empty is not ideal if it was just supposed to be a "quick check beforehand".

nyalldawson replied:
Ok, then why not just fallback to the single feature request in the base method?

m-kuhn replied:
Not sure I understand this. You mean requesting features with LIMIT 1 directly in the `empty()` method?

nyalldawson replied:
Yeah, just like you've done with QgsVectorDataProvider::empty()

m-kuhn replied:
As far as I know, processing feature sources do not (always) send their filter queries to the database, even less in future piped mode. In this case they'll locally evaluate expression or geometry filters.
If we want to know if a filter on a (potentially huge) source is empty we need to iterate and apply the expression to every feature which could lead to a costly loop. I want this method to be cheap by default. And I want to give the developer the required tools and knowledge to take an informed decision if getting reliable information might be more expensive.

nyalldawson replied:
I'm a bit confused here - why would requesting one feature require us to iterate the whole layer? There's no query involved here as far as I can see.

Taking a few steps back, I wonder if an alternative approach would be to flip the method to "hasFeatures", and return the three state enum from that (where 0 = definitely no features)?

m-kuhn replied:
> I'm a bit confused here - why would requesting one feature require us to iterate the whole layer? There's no query involved here as far as I can see.

If we are talking about (filter) algorithms (and not data providers), we'll have to run the algorithm until we find the first feature that passes the filter. If there are no features that pass the filter, we will have run the whole algorithm and consumed the whole data provider to find out that it's empty.

> Taking a few steps back, I wonder if an alternative approach would be to flip the method to "hasFeatures", and return the three state enum from that (where 0 = definitely no features)?

That sounds like an approach, I could live with that

nyalldawson commented on src/core/qgsvectordataprovider.cpp:
```diff
@@ -46,6 +46,20 @@ QString QgsVectorDataProvider::storageType() const
+bool QgsVectorDataProvider::empty() const
+{
+  QgsFeature f;
+  if ( getFeatures().nextFeature( f ) )
```
Use an optimised request here- limit of 1, no attributes, no geometry

nyalldawson commented on src/core/qgsvectorlayer.cpp:
```diff
@@ -2750,6 +2752,11 @@ long QgsVectorLayer::featureCount() const
 
+bool QgsVectorLayer::empty() const
+{
+  return mDataProvider->empty() && ( !mEditBuffer || mEditBuffer->addedFeatures().empty() );
```
Edge case: what about a feature which is added and then deleted in the buffer?

m-kuhn replied:
Give me a performant approach that implements this and I'll do it for ya.

nyalldawson replied:
Left as an exercise for the reader ;)

nyalldawson commented on src/core/qgsvectorlayer.h:
```diff
@@ -921,6 +921,17 @@ class CORE_EXPORT QgsVectorLayer : public QgsMapLayer, public QgsExpressionConte
+    /**
+     * Determines if this vector layer is empty.
+     * A layer is considered empty if either the data provider or
+     * the edit buffer have features in them.
```
" if either the data provider or the edit buffer have features in them" -> "if neither the data provider or the edit buffer have features in them"

m-kuhn replied:
neither ... nor?

nyalldawson replied:
Correct! You win the good grammar award today

nyalldawson commented on src/core/qgsvectorlayer.h:
```diff
@@ -921,6 +921,17 @@ class CORE_EXPORT QgsVectorLayer : public QgsMapLayer, public QgsExpressionConte
+     * Determines if this vector layer is empty.
+     * A layer is considered empty if either the data provider or
+     * the edit buffer have features in them.
+     * This means, in case the data provider contains features which have all
```
This is a bit unclear- it took me a couple of reads to understand. Maybe this is better:

```
\warning A layer is considered empty if neither the data provider or the edit buffer have features in them. This means the returned value may be inaccurate for layers with unsaved changes, e.g. in the case of all features from the data provider having been deleted within the edit buffer"

m-kuhn replied:
`s/e.g./i.e./` and good to go

nyalldawson commented on tests/src/python/providertestbase.py:
```diff
@@ -415,6 +415,21 @@ def testFeatureCount(self):
+            self.assertTrue(self.source.empty())
+            self.source.setSubsetString(None)
+            self.assertFalse(self.source.empty())
+
```
You could also add a test like testTruncate from providertestbase.py - i.e. get an editable layer (if supported by the provider's test), delete all features (by calling truncate), and then test if the layer returns true for isEmpty(). This would directly test the provider's implementation without the subset strings coming into play at all.

m-kuhn committed:
Improve API docs

m-kuhn committed:
Use cheap request

m-kuhn committed:
QgsFeatureSource::hasFeatures() -> Yes, No, Maybe

This method allows to determine if a feature source has content or not
(or maybe).

This method works for "generator" sources as well as for "collection"
sources.

Collection sources like QgsVectorDataProvider which allow to easily and
in a cheap way determine if features are available can return
FeaturesAvailable or NoFeaturesAvailable, generators will return
FeaturesMaybeAvailable.

QgsVectorDataProvider also implements an empty() method.

nyalldawson reviewed (commented):
Don't forget to implement in QgsProcessingFeatureSource too!

m-kuhn commented on src/core/qgsvectorlayer.cpp:
```diff
@@ -2750,6 +2752,14 @@ long QgsVectorLayer::featureCount() const
 
+QgsFeatureSource::FeatureAvailability QgsVectorLayer::hasFeatures() const
+{
+  if ( mDataProvider->empty() && ( !mEditBuffer || mEditBuffer->addedFeatures().empty() ) )
```
You mean if deletedFeatures is not empty? One could have removed all provider features.

nyalldawson replied:
isn't that what I said? :stuck_out_tongue:

m-kuhn replied:
addedFeatures sounded like edit buffer, but if that's what you meant, all good ;)

nyalldawson replied:
AH - I get you. You mean only checking deletedFeatures. Sounds logical!

nyalldawson commented on src/core/qgsvectordataprovider.h:
```diff
@@ -160,6 +160,25 @@ class CORE_EXPORT QgsVectorDataProvider : public QgsDataProvider, public QgsFeat
+     * \since QGIS 3.2
+     * \see empty()
+     */
+    virtual QgsFeatureSource::FeatureAvailability hasFeatures() const override;
```
I think it's redundant since the override keyword was introduced. And now tools like the clang tidy modernize checks automatically strip it out too.

m-kuhn replied:
One could claim that the `override` keyword itself is redundant information too, still I'd never argue to remove that for the technical safety it adds. I'm happy to also keep some "clutter" around for visual clarity (like typenames instead of auto). Anyway, let's go for the clang tidy standard :stuck_out_tongue_closed_eyes:

m-kuhn committed:
Make QgsVectorDataProvider::hasFeatures() final

m-kuhn committed:
Advanced logic for QgsVectorLayer::hasFeatures()

m-kuhn committed:
Fix docstring

m-kuhn commented on src/core/qgsvectordataprovider.h:
```diff
@@ -160,6 +160,25 @@ class CORE_EXPORT QgsVectorDataProvider : public QgsDataProvider, public QgsFeat
+    virtual bool empty() const;
+
+    /**
+     * Will always return FeatureAvailability::FeaturesAvailable or
```
Thanks :/

m-kuhn committed:
Invert logic

because True is not False

m-kuhn committed:
Adjust tests

m-kuhn committed:
Fix test

m-kuhn committed:
Adjust docstrings

m-kuhn commented:
@nyalldawson what's your opinion on this docstring, adjust it to make the test pass or can I get the test to ignore this line?

>   FeatureAvailability: Use "Returns the..." instead of "return ..."

```
    /**
     * Return value for hasFeatures() to determine if a source is empty.
     * It is implemented as a three-value logic, so it can return if
     * there are features available for sure, if there are no features
     * available for sure or if there might be features available but
     * there is no guarantee for this.
     *
     * \since QGIS 3.4
     */
    enum FeatureAvailability
    {
      NoFeaturesAvailable, //!< There are certainly no features available in this source
      FeaturesAvailable, //!< There is at least one feature available in this source
      FeaturesMaybeAvailable //!< There may be features available in this source
    };
```

m-kuhn committed:
Make doctest happy

m-kuhn committed:
Adjust target version

m-kuhn committed:
Fix QgsVectorLayer::hasFeatures() crash when editable

m-kuhn committed:
Add an empty test on truncated layers

m-kuhn committed:
Reset subset string after testing

Wfs test fails otherwise.

m-kuhn commented:
@rouault do you have an idea why the wfs tests are failing here?

It looks as if the dataset is somehow modified for them but I can't see why.

m-kuhn committed:
Isolate test failures with subset string modification

m-kuhn committed:
Workaround to fix WFS provider empty() method

m-kuhn commented on src/providers/wfs/qgswfsprovider.cpp:
```diff
@@ -1187,6 +1187,27 @@ QString QgsWFSProvider::translateMetadataValue( const QString &mdKey, const QVar
+  // Remove this method (and default to the QgsVectorDataProvider one)
+  // once this is fixed
+#if 0
+  request.setLimit( 1 );
```
@rouault any idea why the WFS provider returns an empty result set when using a combination of a subsetString on the provider and a request with a limit set?

m-kuhn committed:
Change docstring order

nyalldawson commented:
@m-kuhn I was about to tag this "merge after thaw", but noticed that this still needs addressing:

> Don't forget to implement in QgsProcessingFeatureSource too!

So now you get a "requires changes" tag instead :stuck_out_tongue:

m-kuhn commented:
I can add a default implementation inside `QgsProcessingFeatureSource::hasFeatures()` that returns `FeaturesMaybeAvailable` if you think the default one on `QgsFeatureSource` doing the same is not sufficient ;P

nyalldawson commented:
But if the underlying source is a postgres provider, you'd want the optimised call happening, right?

m-kuhn committed:
Add minimal smartness for QgsProcessingFeatureSource::hasFeatures

m-kuhn commented:
Better?

nyalldawson commented on src/core/processing/qgsprocessingutils.cpp:
```diff
@@ -721,6 +721,14 @@ QgsFeatureIterator QgsProcessingFeatureSource::getFeatures( const QgsFeatureRequ
 
+QgsFeatureSource::FeatureAvailability QgsProcessingFeatureSource::hasFeatures() const
+{
+  if ( !mTransformErrorCallback && !mInvalidGeometryCallback && mInvalidGeometryCheck == QgsFeatureRequest::GeometryNoCheck )
```
We should never be fetching geometry during a has features check, so this method should just be

    return mSource->hasFeatures();

nyalldawson replied:
Ah, I see where you're going here.

The two callback checks can be skipped - they can't be used to filter features, just handle when they are encountered. 

An optimised method could be:

    FeatureAvailability sourceAvailability = mSource->hasFeatures();
    if ( sourceAvailability == NoFeaturesAvailable )
         return NoFeaturesAvailable; // never going to be features if underlying source has no features
    else if ( mInvalidGeometryCheck == QgsFeatureRequest::GeometryNoCheck )
         return sourceAvailability;
    else
         // we don't know... source has features, but these may be filitered out by invalid geometry check
         return FeaturesMaybeAvailable;

m-kuhn replied:
Sounds good, thanks

m-kuhn committed:
Update to QgsProcessingFeatureSource::hasFeatures
